pkg/apis/core/v1alpha1/filewatch_types.go

```diff
@@ -53,6 +53,12 @@ type FileWatchList struct {
 
 // FileWatchSpec defines the desired state of FileWatch
 type FileWatchSpec struct {
+	RootPath string `json:"rootPath"`
+
+	Paths []string `json:"matchPatterns,omitempty"`
+
+	// +optional
+	IgnorePatterns []string `json:"ignorePatterns,omitempty"`
```

Comment:
i feel like we often need to model sets of ignore patterns with different roots? e.g., a .dockerignore and a tiltfile-generated ignores. is your plan to normalize them to the same root path?

Reply:
I originally had the impression that the base path was always the same, perhaps something like this makes more sense?

```go
type PatternSet struct {
   BasePath string
   Patterns []string
}

type FileWatchSpec struct {
   Watches []PatternSet
   Ignores []PatternSet
}
```

Then the controller could create a set of absolute path patterns from the watches for the actual FS watcher logic and a `CompositeMatcher` from the various ignores?

Taking things up one level - I _think_ it makes sense for each `FileWatch` to be self-contained, but I'm still not 100% sure what/how the flow for propagating global ignores should be. (cc @ellenkorbes)

Reply:
ya, that should work

Reply:
fwiw, most build systems don't support arbitrary Watch/Ignore patterns, because it can lead to pathological performance characteristics if you're not careful. 

Most things in Tilt let you watch a directory (recursively) and then set up patterns to ignore (with some constraints on those patterns), but even this has some (intentional) bugs to avoid perf regressions, see https://github.com/moby/moby/issues/30018

Reply:
Okay, maybe this which avoids the terminology ambiguity and some repetition for simplistic cases:
```go
type WatchSet struct {
   BasePath string
   Paths []string
   IgnorePatterns []string
}

type FileWatchSpec {
   WatchSets []WatchSet
}
```

```yaml
spec:
  watchSets:
    - basePath: "/a"
      paths: ["./b", "c"]
      ignores: ["c/d"]
    - basePath: "/a/b"
      ignores: ["zzz"]
```

Same logic applies - create absolute paths for all the watched paths and a `CompositeMatcher` from the various ignores

Reply:
New version uses this, I like it so far? Will update GDoc